### ISSUE TITLE
Incorrectly reporting usage errors for decorated functions

### ISSUE THREAD

**Issue by abingham** [microsoft/pyright#9465]
## Environment data

-   Pylance version: v2024.11.2
-   OS and version: macOS 14.6.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.5

## Code Snippet

```python
def dec(f):
    def wrapper():
        return f(41)
    return wrapper

@dec
def fnord(x):
    print(x + 1)

fnord()  # <--- Pylance incorrectly flags this as an error
```

## Repro Steps

1. Open this code in vscode with the pylance extension enabled.

## Expected behavior

Pylance doesn't flag any errors with the code.

## Actual behavior

Pylance claims that 'Argument missing for parameter "x"' at the call to `fnord()`.

Since the wrapper around fnord produced by the decorator does not require any arguments, pylance seems to be incorrect about my usage here. Indeed, I can run this program with no problems.

## Logs

Not entirely sure which log you want, so here's the log from my "Python Language Server" output tab when I trigger re-analysis of the offending file:

```
2024-11-14 11:46:36.341 [info] (83514) [BG(1)] InlayHints range 0:0 - 10:0 at file:///Users/austin/repos/cosmic-ray/deleteme.py ...
2024-11-14 11:46:36.341 [info] (83514) [BG(1)]   parsing: file:///Users/austin/repos/cosmic-ray/deleteme.py (1ms)
2024-11-14 11:46:36.342 [info] (83514) [BG(1)]   binding: file:///Users/austin/repos/cosmic-ray/deleteme.py (0ms)
2024-11-14 11:46:36.346 [info] (83514) [BG(1)] InlayHints range 0:0 - 10:0 at file:///Users/austin/repos/cosmic-ray/deleteme.py (5ms)
2024-11-14 11:46:36.516 [info] (83514) [FG] parsing: file:///Users/austin/repos/cosmic-ray/deleteme.py (1ms)
2024-11-14 11:46:36.516 [info] (83514) [FG] binding: file:///Users/austin/repos/cosmic-ray/deleteme.py (0ms)
2024-11-14 11:46:36.566 [info] (83514) [BG(1)] indexing: file:///Users/austin/repos/cosmic-ray/deleteme.py [found 2] (0ms)
2024-11-14 11:46:36.566 [info] (83514) Workspace indexing done: file:///Users/austin/repos/cosmic-ray/deleteme.py
2024-11-14 11:46:36.566 [info] [Info  - 11:46:36] (83514) Heap stats: total_memory_size=32768MB, total_free_size=398MB, total_heap_size=46MB, used_heap_size=42MB, cross_worker_used_heap_size=42MB, total_physical_size=45MB, total_available_size=4053MB, heap_size_limit=4096MB
2024-11-14 11:46:36.567 [info] (83514) [BG(1)] analyzing: file:///Users/austin/repos/cosmic-ray/deleteme.py ...
2024-11-14 11:46:36.567 [info] (83514) [BG(1)]   checking: file:///Users/austin/repos/cosmic-ray/deleteme.py (1ms)
2024-11-14 11:46:36.567 [info] (83514) [BG(1)] analyzing: file:///Users/austin/repos/cosmic-ray/deleteme.py (1ms)
2024-11-14 11:46:36.575 [info] (83514) [BG(1)] SemanticTokens delta previousResultId:1731581186702 at file:///Users/austin/repos/cosmic-ray/deleteme.py (0ms)
2024-11-14 11:46:37.116 [info] (83514) [BG(1)] InlayHints range 0:0 - 10:0 at file:///Users/austin/repos/cosmic-ray/deleteme.py (1ms)
2024-11-14 11:46:37.996 [info] [Info  - 11:46:37] (83514) SourceFile: Received fs event 'change' for path '/Users/austin/repos/cosmic-ray/deleteme.py'
2024-11-14 11:46:37.997 [info] [Info  - 11:46:37] (83514) SourceFile: Received fs event 'add' for path '/Users/austin/repos/cosmic-ray/.mypy_cache/3.12/@plugins_snapshot.json'
2024-11-14 11:46:37.997 [info] [Info  - 11:46:37] (83514) SourceFile: Received fs event 'add' for path '/Users/austin/repos/cosmic-ray/.mypy_cache/3.12/deleteme.meta.json'
2024-11-14 11:46:37.997 [info] [Info  - 11:46:37] (83514) SourceFile: Received fs event 'change' for path '/Users/austin/repos/cosmic-ray/.mypy_cache/3.12/deleteme.meta.json.f1030e110ec9c26f'
2024-11-14 11:46:37.997 [info] [Info  - 11:46:37] (83514) SourceFile: Received fs event 'change' for path '/Users/austin/repos/cosmic-ray/.mypy_cache/3.12/@plugins_snapshot.json.d7fce395cdf1dccb'
2024-11-14 11:46:38.003 [info] (83514) [BG(1)] indexing: file:///Users/austin/repos/cosmic-ray/deleteme.py [found 2] (0ms)
2024-11-14 11:46:38.003 [info] (83514) Workspace indexing done: file:///Users/austin/repos/cosmic-ray/deleteme.py
2024-11-14 11:46:38.003 [info] [Info  - 11:46:38] (83514) Heap stats: total_memory_size=32768MB, total_free_size=310MB, total_heap_size=46MB, used_heap_size=43MB, cross_worker_used_heap_size=43MB, total_physical_size=45MB, total_available_size=4053MB, heap_size_limit=4096MB
2024-11-14 11:46:38.323 [info] [Info  - 11:46:38] (83514) SourceFile: Received fs event 'change' for path '/Users/austin/repos/cosmic-ray/tests/.pytest_cache/v/cache/stepwise'

```

**Comment by erictraut** [comment]
Pyright's current behavior is intended, so I don't consider this a bug.

When a decorator has no type annotations, it is assumed to not modify the signature of the decorated function. This is the safest assumption when no type information is available. If the decorator is intended to modify the decorated function's signature, you can indicate as much by providing (even minimal) type annotations.

```python
def dec(f):
    def wrapper() -> None: # Annotation indicates return type of None in this case
        return f(41)

    return wrapper
```

Or you can provide more complete type annotations if you'd like to further document the signature of your decorator and provide the type checker with more information.

```python
def dec(f: Callable[[int], None]) -> Callable[[], None]:
    def wrapper() -> None:
        return f(41)

    return wrapper
```
